test/lib/ansible_test/_data/sanity/code-smell/runtime.py

```diff
@@ -0,0 +1,149 @@
+#!/usr/bin/env python
+"""Schema validation of ansible-base's ansible_builtin_runtime.yml and collection's meta/runtime.yml"""
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+import os
+import re
+import sys
+import yaml
+
+from voluptuous import Any, MultipleInvalid, PREVENT_EXTRA
+from voluptuous import Required, Schema
+from voluptuous.humanize import humanize_error
+
+from ansible.module_utils.six import string_types
+
+
+def main():
+    """Validate runtime"""
+    ansible_builtin_runtime = 'lib/ansible/config/ansible_builtin_runtime.yml'
+    collection_legacy_file = 'meta/routing.yml'
+    collection_runtime_file = 'meta/runtime.yml'
+    path = ''
+
+    # Check if the correct routing files can be found
+
+    if os.path.isfile(ansible_builtin_runtime):
+        # We are in side an ansible-base checkout, so validate ansible_builtin_runtime
+        path = ansible_builtin_runtime
+    elif os.path.isfile(collection_runtime_file):
+        # We are in a collection, so validate meta/routing.yml
+        path = collection_runtime_file
+
+    if os.path.isfile(collection_legacy_file):
+        # Collection contains old name, renamed in #67684
+        print('%s:%d:%d: %s' % (collection_legacy_file, 0, 0, ("Should be called '%s'" % collection_runtime_file)))
+
+    if not path:
+        # In a collection, though no meta/routing
+        sys.exit(0)
+
+    try:
+        with open(path, 'r') as f_path:
+            routing = yaml.safe_load(f_path)
+    except yaml.error.MarkedYAMLError as ex:
+        print('%s:%d:%d: YAML load failed: %s' % (path, ex.context_mark.line +
+                                                  1, ex.context_mark.column + 1, re.sub(r'\s+', ' ', str(ex))))
+        sys.exit()
+    except Exception as ex:  # pylint: disable=broad-except
+        print('%s:%d:%d: YAML load failed: %s' %
+              (path, 0, 0, re.sub(r'\s+', ' ', str(ex))))
+        sys.exit()
+
+    # Updates to schema MUST also be reflected in the documentation
+    # ~https://docs.ansible.com/ansible/devel/dev_guide/developing_collections.html
+
+    # plugin_routing schema
+
+    deprecation_schema = Any(Schema(
```

Comment:
The required vs options here don't seem to align with what's in ansible-base.
Apparently, ansible-base has some default text.
@felixfontein suggested that maybe one of `removal_date` or `removal_version` be required, and that `warning_test` be optional.

Reply:
I would make `removal_date` and `removal_version` mutually exclusive and required (i.e. precisely one of them must be there), and make `warning_test` optional. This is compatible to my proposed change here: #69798

(Allowing to specify neither `removal_date` nor `removal_version` is not a good idea IMO - if someone really wants to do it, they can still add ignore.txt entries. But it should not be encouraged - users want to know how long they can still use something.)